datreant/backends/core.py

```diff
@@ -316,3 +329,139 @@ def _close(self):
         self._unlock(self.fd)
         self.fdlock = None
         self._close_fd()
+
+
+class FileSerial(File):
+    """File object base class for serialization formats, such as JSON.
+
+    """
+    def _open_file_r(self):
+        return open(self.filename, 'r')
+
+    def _open_file_w(self):
+        return open(self.filename, 'w')
+
+    def read_file(self):
+        """Return deserialized representation of file.
+
+        """
+        self._apply_shared_lock()
+
+        self.handle = self._open_file_r()
+        out = self._deserialize(self.handle)
+        self.handle.close()
+
+        self._release_lock()
+
+    @staticmethod
+    def _read(func):
+        """Decorator for applying shared lock on file.
+
+        Applying this decorator to a method will ensure a shared lock is
+        obtained before that method is executed. It also ensures that the lock
+        is removed after the method returns.
+
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            if self.fdlock:
+                out = func(self, *args, **kwargs)
+            else:
+                self._apply_shared_lock()
+                try:
+                    out = func(self, *args, **kwargs)
+                finally:
+                    self._release_lock()
+            return out
+
+        return inner
+
+    @staticmethod
+    def _write(func):
+        """Decorator for applying an exclusive lock on file.
+
+        Applying this decorator to a method will ensure an exclusive lock is
+        obtained before that method is executed. It also ensures that the lock
+        is removed after the method returns.
+
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            if self.fdlock == 'exclusive':
+                out = func(self, *args, **kwargs)
+            else:
+                self._apply_exclusive_lock()
+                try:
+                    out = func(self, *args, **kwargs)
+                finally:
+                    self._release_lock()
+            return out
+
+        return inner
+
+    @staticmethod
+    def _pull_push(func):
+        """Decorator for deserializing the contents of file, then reserializing.
+
+        Applying this decorator to a method will ensure the latest version of
+        the data is obtained before that method is executed. It also ensures
+        that changes to the data are written to the file after the method
+        returns.
+
+        This decorator doesn't do any file locking, but should be used after
```

Comment:
I like it better when you decorate the push pull functions. It keeps them nice and short plus the locking is stated very obvious when one looks at the definition of the function.